### ISSUE TITLE
cmd/compile: cannot use s (type [][]Typ) as type []go.shape.[]int_0 in assignment

### ISSUE THREAD

**Issue by YueMengGu** [golang/go#55964]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18.6 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/xxx/.cache/go-build"
GOENV="/Users/xxx/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/xxx/Coding/go/pkg/mod"
GONOPROXY="*.everphoto.cn,git.smartisan.com"
GONOSUMDB="*.everphoto.cn,git.smartisan.com"
GOOS="darwin"
GOPATH="/Users/xxx/Coding/go"
GOPRIVATE="*.everphoto.cn,git.smartisan.com"
GOPROXY="https://goproxy.cn,https://proxy.golang.org,direct"
GOROOT="/Users/xxx/Coding/github/go"
GOSUMDB="sum.golang.google.cn"
GOTMPDIR=""
GOTOOLDIR="/Users/xxx/Coding/github/go/pkg/tool/darwin_amd64"
GOVCS=""
GOVERSION="go1.18.6"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/xxx/Coding/tmp/testgo/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch x86_64 -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/77/qcdftwkd6jb6945_b2k6n60h0000gp/T/go-build1909810451=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

compile the code with `go build`
```go
package main

type iter[T any] struct {
	s []T
}

func StealSlice[T any](s []T) *iter[T] {
	return &iter[T]{s}
}

func Flatten[T any](s [][]T) {
	StealSlice(s)
}

func main() {
	type Typ int
	st := [][]Typ{{11}, {22, 33}}
	Flatten(st)
}
```

### What did you expect to see?
compile succeed

### What did you see instead?
./main.go:18:9: cannot use s (type [][]Status) as type []go.shape.[]int_0 in assignment

if add `//go:noinline` upper `Flatten`, compile succeed

**Comment by mdempsky** [comment]
This appears fixed in Go 1.19. I don't think we're going to backport a fix for Go 1.18. Sorry.

**Comment by YueMengGu** [comment]
> This appears fixed in Go 1.19 (https://go.dev/play/p/oPCFs0AZEie) and tip (https://go.dev/play/p/oPCFs0AZEie?v=gotip). I don't think we're going to backport a fix for Go 1.18. Sorry.

Thanks.

Another two questions,
1. What‘s the root cause? 
2. Why not to backport, whether it is too compliex or something else
